### PR TITLE
fix(schema-compat): prevent duplicate optional tool schemas

### PR DESCRIPTION
**Pull request by NikAiyer**
We found this while switching an active Mastra Code thread from Claude Fable 5 to `gpt-5.6-sol`. The OpenAI turn immediately failed with “The model stopped because it reached its maximum output length,” and `gpt-5.5` produced the same result. The thread itself was not unusually large, and direct requests to both models worked normally.

Capturing and replaying the actual request showed that the failure depended on the attached MCP tools. Mastra Code sent 159 tools with Notion MCP enabled; disabling Notion reduced that to 118 tools and made the same `gpt-5.6-sol` turn succeed. Bisecting the Notion tools isolated `notion_notion-query-meeting-notes` as the trigger.

That tool exposes a recursive meeting filter supporting nested `and`/`or` groups, title filters, attendee filters, and several date-filter variants. Its input schema contains deeply nested optional object definitions. OpenAI schema compatibility made those properties nullable while leaving their object-specific keywords on the containing property, so the same recursive subtree appeared both outside and inside its `anyOf` object branch. This expanded the meeting-notes tool schema to roughly 69 KB.

The Codex backend rejected that schema before inference but reported the rejection as `response.incomplete` with `reason: max_output_tokens`, zero input tokens, and zero output tokens. That misleading response is why the issue initially looked like a model output-budget problem rather than a tool-schema problem.

Before:

```json
{
  "properties": { "field": {} },
  "anyOf": [
    { "type": "object", "properties": { "field": {} } },
    { "type": "null" }
  ]
}
```

After:

```json
{
  "anyOf": [
    { "type": "object", "properties": { "field": {} } },
    { "type": "null" }
  ]
}
```

This keeps object and array-specific keywords only in their matching union branch while preserving shared metadata on the containing property. It reduces the affected meeting-notes schema to roughly 22 KB without disabling or special-casing Notion, and fixes the general conversion behavior for other deeply nested MCP schemas.

Regression coverage verifies recursive definitions are emitted once and nested optional schemas grow linearly. The change was also verified with package lint and builds, the Mastra Code SDK build, and a replay of the original 159-tool Notion MCP request against `gpt-5.6-sol`, which completed normally after the fix.

## ELI5

This PR prevents optional tool schemas from being copied multiple times during OpenAI conversion. Smaller schemas prevent valid requests from failing with misleading token-limit errors.

## Changes

- Keeps object- and array-specific keywords inside their matching nullable `anyOf` branches.
- Preserves metadata, `enum`, `const`, and composition constraints.
- Adds regression tests for recursive schemas, null handling, and linear schema growth.
- Updates strict-schema assertions and affected OpenAI recordings.
- Adds a patch changeset for `@mastra/schema-compat`.
- Excludes `openai-organization` and `openai-project` from recorded headers.
- Expands the model compatibility matrix and replaces the delisted Gemini model.

## Validation

- Ran linting and builds.
- Built the Mastra Code SDK.
- Replayed the original 159-tool Notion MCP request against `gpt-5.6-sol`.
- Reduced the affected schema from approximately 69 KB to approximately 22 KB.

### CODE REVIEW

File: packages/schema-compat/src/provider-compats/openai.ts [hunk 1]
```diff
@@ -304,17 +306,33 @@ export class OpenAISchemaCompatLayer extends SchemaCompatLayer {
...
-              delete prop.type;
-              prop.anyOf = [
-                {
-                  ...propSchema,
-                  type: originalType,
-                },
-                { type: 'null' },
-              ];
+              const keywords = typeSpecificKeywords(originalType);
+              if (keywords.length > 0) {
```
[the remark this answers is not in the document]
Comment (NikAiyer) [reply packages/schema-compat/src/provider-compats/openai.ts:310]: AI says: Good catch. I moved `enum` and `const` into the non-null object/array branch and added regression coverage confirming the generated null branch remains unconstrained and validates through the null-to-undefined path. I also updated the shared Groq snapshot and verified the full schema-compat suite (904 passed, 9 skipped), lint, and build. Fixed in 33ec4cbf85. @coderabbitai

Mastra Code

File: mastracode/tui/e2e/tui/openai-strict-schema.ts [hunk 2]
```diff
@@ -107,14 +107,26 @@ export const openaiStrictSchemaScenario = {
...
     check(
-      JSON.stringify(requiredKeys(nested)) === JSON.stringify(sortedKeys(nested.properties)),
-      `Expected all nested properties to be required, received required=${JSON.stringify(nested.required)} properties=${JSON.stringify(sortedKeys(nested.properties))}`,
+      !('properties' in nested),
+      `Expected nested properties to be branch-local, received ${JSON.stringify(nested)}`,
+    );
+    check(nestedObjectBranch.additionalProperties === false, 'Expected nested additionalProperties false');
+    check(
+      JSON.stringify(requiredKeys(nestedObjectBranch)) === JSON.stringify(sortedKeys(nestedObjectBranch.properties)),
+      `Expected all nested properties to be required, received required=${JSON.stringify(nestedObjectBranch.required)} properties=${JSON.stringify(sortedKeys(nestedObjectBranch.properties))}`,
```
[the remark this answers is not in the document]
Comment (NikAiyer) [reply mastracode/tui/e2e/tui/openai-strict-schema.ts]: AI says: @coderabbitai Addressed in d15e161006. The E2E now explicitly asserts that the nested object branch declares exactly `count` and `enabled`, while retaining the existing check that every declared property is required. Focused Mastra Code E2E, TypeScript check, and lint pass.

File: packages/core/src/tools/tool-builder/builder.e2e.test.ts [hunk 3]
```diff
@@ -327,27 +327,27 @@ const modelsToTestV1 = [
   // openrouter('anthropic/claude-3.7-sonnet'),
   // openrouter('anthropic/claude-sonnet-4.5'),
   openrouter('anthropic/claude-haiku-4.5'),
-  // openrouter('openai/gpt-4o-mini'),
-  // openrouter('openai/gpt-4.1-mini'),
-  // openrouter_v5('openai/o3-mini'),
-  // openai('o3-mini'),
+  openrouter('openai/gpt-4o-mini'),
```
[the remark this answers is not in the document]
Comment (NikAiyer) [reply packages/core/src/tools/tool-builder/builder.e2e.test.ts:329]: AI says: These tests do not make live API calls in CI. This file uses `@internal/llm-recorder` (`createGatewayMock` at the top of the file): in the default `auto` mode, requests are replayed from the checked-in recordings in `packages/core/__recordings__/core-src-tools-tool-builder-builder.e2e.json`, and `setupDummyApiKeys` injects placeholder keys (e.g. `sk-dummy-for-replay-mode`) so no real provider credentials are even present. Recordings for every newly enabled model were captured in this PR, and a strict-replay run (`LLM_TEST_MODE=replay`, zero network) passes 50/50. Live spend only occurs when a developer explicitly re-records with `UPDATE_RECORDINGS=true`, which already is the opt-in gate this finding asks for.

File: packages/core/__recordings__/core-src-background-tasks-background-tasks.e2e/test-5e774f5f.json [hunk 4]
```diff
@@ -195,68 +146,46 @@
...
-          "openai-organization": "mastra-qb2kpb",
-          "openai-processing-ms": "150",
-          "openai-project": "proj_Cvw1JiXoxHTeEWgnnK3t9AYG",
+          "date": "Wed, 02 Sep 2026 22:24:06 GMT",
+          "openai-organization": "personal-1smmvi",
+          "openai-processing-ms": "237",
+          "openai-project": "proj_i5xepL7AVq5Id7m6q2HKx0cV",
           "openai-version": "2020-10-01",
           "server": "cloudflare",
           "strict-transport-security": "max-age=31536000; includeSubDomains; preload",
```
[the remark this answers is not in the document]
Comment (NikAiyer) [reply packages/core/__recordings__/core-src-background-tasks-background-tasks.e2e/test-5e774f5f.json:156]: AI says: Fixed in 5de14d6f00. Added `openai-organization` and `openai-project` to the llm-recorder `SKIP_HEADERS` list so future recordings never capture them, and scrubbed the existing values (188 occurrences) from every recording touched by this PR. Note these are account identifiers rather than credentials, and the same headers already exist in pre-existing recordings on main that this PR doesn't touch (e.g. `core-src-tools-provider-tools.e2e.json`, the durable-agent recordings) — those can be cleaned up on their next re-record now that the recorder strips them at capture time. Strict replay of all affected suites passes (95 tests), and llm-recorder's own tests and lint pass.

**Review by TylerBarnes — APPROVED** [review]
The branch-local nullable schema approach looks sound. Reviewed the conversion and runtime traversal, and verified package tests, build, and typecheck. Also compared base/PR schemas against the OpenAI API and Codex backend; the investigated edge cases showed no new provider failures. No blocking findings.
